formats/roam/roam.ts

```diff
@@ -0,0 +1,327 @@
+  };
+
+async function jsonToMarkdown(graphFolder:string, attachmentsFolder:string, downloadAttachments:boolean,json: JsonObject, indent: string = '', isChild: boolean = false ): Promise<string> {
+    let markdown = '';
+
+    if (json.string) {
+        let prefix = '';
+        if (json.heading) {
+            prefix = '#'.repeat(json.heading) + ' ';
+        }
+        markdown += `${isChild ? indent + '* ' : indent}${prefix}${(await roamMarkupScrubber(graphFolder, attachmentsFolder, json.string, downloadAttachments))}\n`;
+    }
+
+    if (json.children) {
+        for (const child of json.children) {
+            markdown += await jsonToMarkdown(graphFolder, attachmentsFolder, downloadAttachments, child, indent + '  ', true);
+        }
+    }
+
+    return markdown;
+}
+
+export async function importRoamJson (importer:RoamJSONImporter, progress: ProgressReporter, files:PickedFile[], folder:TFolder, downloadAttachments:boolean = true) {
+    for (let file of files) {
+        const graphName = sanitizeFileName(file.basename);
+        const graphFolder = path.join(folder.path, graphName);
+		const attachmentsFolder = path.join(folder.path, graphName, "Attachments")
+        // create the base graph folders
+        await importer.createFolders(graphFolder)
+		await importer.createFolders(attachmentsFolder)
+
+		// read the graph
+		// TODO is this async?
+        const data = fs.readFileSync(file.filepath, "utf8")
+        const allPages = JSON.parse(data) as RoamPage[]
+
+        // PRE-PROCESS: map the blocks for easy lookup //
+        const [blockLocations, toPostProcess] = preprocess(allPages)
+		
+        // WRITE-PROCESS: create the actual pages //
+        for (let index in allPages) {
+            const pageData = allPages[index]
+
+			const pageName = convertDateString(sanitizeFileNameKeepPath(pageData.title), userDNPFormat)
+            const filename =  path.join(graphFolder, `${pageName}.md`)
+            // convert json to nested markdown
+
+			const markdownOutput = await jsonToMarkdown(graphFolder, attachmentsFolder, downloadAttachments, pageData);
+            
+            try {
+				//create folders for nested pages [[some/nested/subfolder/page]]
+                await importer.createFolders(path.dirname(filename))
+				const existingFile = app.vault.getAbstractFileByPath(filename);
+				if (existingFile) {
+					await app.vault.adapter.write(existingFile.path, markdownOutput);
+					// console.log("Markdown replaced in existing file:", existingFile.path);
+				} else {
+					const newFile = await app.vault.create(filename, markdownOutput);
+					// console.log("Markdown saved to new file:", newFile.path);
+				}
+				progress.reportNoteSuccess(filename);
+			} catch (error) {
+				console.error("Error saving Markdown to file:", filename, error);
+				progress.reportFailed(pageName);
+			}
+        }
+
+		// POST-PROCESS: fix block refs //
+		async function modifySourceBlockString(sourceBlockUID:string) {
```

Comment:
This function is awfully bad for performance as it keeps reading files, blowing it up into lines, modifying a single line and re-save it to disk. Can we prepare a list of block changes per target file and do a single read-write (using `vault.process`) per file instead?